MIR borrowck: error message confuses locals and temporaries

davidtwco opened the pull request:
Fixes #46471 and fixes #46472 (see [this Gitter comment](https://gitter.im/rust-impl-period/WG-compiler-nll?at=5a2d5cb53ae2aa6b3facf0c2)).

r? @arielb1

davidtwco commented:
This isn't quite there yet, the error output is approximately correct, but the spans are a character to the left of what they should be to match exactly. I'm also not sure that the condition I'm checking is what I should be using, but it works.

davidtwco commented:
I've fixed issues referenced in my previous comment and also fixed #46472 as mentioned in [this Gitter comment](https://gitter.im/rust-impl-period/WG-compiler-nll?at=5a2d5cb53ae2aa6b3facf0c2).

I think the function might be getting a bit a unwieldly since it's handling a bunch of cases. I've adjusted tests to allow the handful of places where spans are off by a character.

arielb1 commented on src/librustc_mir/borrow_check/mod.rs:
```diff
@@ -2373,24 +2371,56 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
-        err.emit();
+        match &self.describe_place(place) {
+            Some(description) => {
+                let mut err = self.tcx.path_does_not_live_long_enough(
```
This doesn't handle the case where a local needs to live until an `end_span`, aka
```Rust
let x = {
    let y = 0;
    &y
};
```

davidtwco replied:
This should now be being handled.

arielb1 commented:
Nice!

@bors r+ p=1

This should go before #46558, which changes the same code.

bors commented:
:pushpin: Commit e1723d2 has been approved by `arielb1`

<!-- @bors r=arielb1 e1723d2a09322af5df3d4e49a7220ab62fca0da9 -->

arielb1 commented:
@bors p=0

This hasn't passed travis yet

kennytm commented:
@bors r-

Travis failed.

```
[00:57:14] failures:
[00:57:14]     [ui] ui/nll/capture-ref-in-struct.rs
[00:57:14]     [ui] ui/nll/closure-requirements/escape-argument.rs
[00:57:14]     [ui] ui/nll/closure-requirements/escape-upvar-nested.rs
[00:57:14]     [ui] ui/nll/closure-requirements/escape-upvar-ref.rs
```

davidtwco commented:
I've fixed the previous issue where a local might need to live until an `end_span`. I've not looked at the other tests which are failing in the Travis build - do these just need updated to have the improved error messages or is this an issue with this PR @arielb1?

davidtwco commented:
I've updated the existing tests to use the new error messages. This should now pass Travis.

arielb1 commented on src/librustc_mir/borrow_check/mod.rs:
```diff
@@ -2373,23 +2371,71 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
+            // See test ui/issue-46471.rs for this case.
+            RegionKind::ReEarlyBound(_) | RegionKind::ReFree(_) => proper_span,
+            // See test compile-fail/issue-36082.rs for this case.
+            _ if name == "borrowed value" => span,
```
nit: use  `if !with_name`

arielb1 commented on src/librustc_mir/borrow_check/mod.rs:
```diff
@@ -2373,23 +2371,71 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
+        };
+
+        let mut err = self.tcx.path_does_not_live_long_enough(err_span, &name, Origin::Mir);
+        match borrow.region {
```
There looks to be quite some connection between the span and the error message, so I would prefer to have 4 separate functions - 1 for each distinct error messagee

kennytm commented:
<details><summary>Travis is still failing.</summary>

```
[00:59:30] ---- [compile-fail] compile-fail/borrowck/borrowck-local-borrow-outlives-fn.rs stdout ----
[00:59:30] 	
[00:59:30] error in revision `mir`: /checkout/src/test/compile-fail/borrowck/borrowck-local-borrow-outlives-fn.rs:15: unexpected error: '15:5: 15:7: `x` does not live long enough [E0597]'
[00:59:30] 
[00:59:30] error in revision `mir`: /checkout/src/test/compile-fail/borrowck/borrowck-local-borrow-outlives-fn.rs:16: expected error not found: borrowed value does not live long enough
[00:59:30] 
[00:59:30] error in revision `mir`: 1 unexpected errors found, 1 expected errors not found
...
[00:59:30] ---- [compile-fail] compile-fail/borrowck/borrowck-local-borrow-with-panic-outlives-fn.rs stdout ----
[00:59:30] 	
[00:59:30] error in revision `mir`: /checkout/src/test/compile-fail/borrowck/borrowck-local-borrow-with-panic-outlives-fn.rs:16: unexpected error: '16:15: 16:23: `z.1` does not live long enough [E0597]'
[00:59:30] 
[00:59:30] error in revision `mir`: /checkout/src/test/compile-fail/borrowck/borrowck-local-borrow-with-panic-outlives-fn.rs:18: expected error not found: [E0597]
[00:59:30] 
[00:59:30] error in revision `mir`: 1 unexpected errors found, 1 expected errors not found
...
[00:59:30] ---- [compile-fail] compile-fail/region-borrow-params-issue-29793-big.rs stdout ----
[00:59:30] 	
[00:59:30] error in revision `mir`: /checkout/src/test/compile-fail/region-borrow-params-issue-29793-big.rs:84: unexpected error: '84:5: 84:6: `x` does not live long enough [E0597]'
[00:59:30] 
[00:59:30] error in revision `mir`: /checkout/src/test/compile-fail/region-borrow-params-issue-29793-big.rs:84: unexpected error: '84:5: 84:6: `y` does not live long enough [E0597]'
[00:59:30] 
[00:59:30] error in revision `mir`: /checkout/src/test/compile-fail/region-borrow-params-issue-29793-big.rs:84: expected error not found: borrowed value does not live long enough
[00:59:30] 
[00:59:30] error in revision `mir`: /checkout/src/test/compile-fail/region-borrow-params-issue-29793-big.rs:84: expected error not found: borrowed value does not live long enough
[00:59:30] 
[00:59:30] error in revision `mir`: 2 unexpected errors found, 2 expected errors not found
```

</details>

If nothing goes wrong, #46558 will be merged soon, so you'll likely need to rebase.

arielb1 commented:
@kennytm 

We're working on it.

bors commented:
:umbrella: The latest upstream changes (presumably #46558) made this pull request unmergeable. Please resolve the merge conflicts.

davidtwco commented:
All previous issues should be resolved and it's now rebased on top of the current master (or as of an hour ago).

davidtwco commented:
Alright, fifth time lucky. I've fixed the indentation and other requested changes from Gitter and updated the failing tests from the recent Travis run.

arielb1 commented:
That's cool!

@bors r+

bors commented:
:pushpin: Commit 3dbc11b has been approved by `arielb1`

<!-- @bors r=arielb1 3dbc11bc6322a8d337ccdf79464e547abafe7f00 -->

bors commented:
:hourglass: Testing commit 3dbc11bc6322a8d337ccdf79464e547abafe7f00 with merge 724500626bcc37db1e271a3c344550fcd4238148...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.5670), [status-travis](https://travis-ci.org/rust-lang/rust/builds/315199383?utm_source=github_status&utm_medium=notification)
Approved by: arielb1
Pushing 724500626bcc37db1e271a3c344550fcd4238148 to master...
